modMcf/src/org/aion/mcf/db/AbstractRepositoryCache.java

```diff
@@ -60,130 +61,203 @@
+    public void loadAccountState(Address address, Map<Address, AccountState> accounts,
             Map<Address, IContractDetails<DataWord>> details) {
+        lockAccounts.readLock().lock();
+        lockDetails.readLock().lock();
+
+        try {
+            // check if the account is cached locally
+            AccountState accountState = this.cachedAccounts.get(address);
+            IContractDetails<DataWord> contractDetails = this.cachedDetails.get(address);
+
+            // when account not cached load from repository
+            if (accountState == null) {
+                // load directly to the caches given as parameters
+                repository.loadAccountState(address, accounts, details);
+            } else {
+                // copy the objects if they were cached locally
+                accounts.put(address, new AccountState(accountState));
+                details.put(address, new ContractDetailsCacheImpl(contractDetails));
+            }
+        } finally {
+            lockAccounts.readLock().unlock();
+            lockDetails.readLock().unlock();
+        }
+    }
 
-        // check if the account is cached locally
-        AccountState accountState = this.cachedAccounts.get(address);
-        IContractDetails<DataWord> contractDetails = this.cachedDetails.get(address);
-
-        // when account not cached load from repository
-        if (accountState == null) {
-            // load directly to the caches given as parameters
-            repository.loadAccountState(address, accounts, details);
-        } else {
-            // copy the objects if they were cached locally
-            accounts.put(address, new AccountState(accountState));
-            details.put(address, new ContractDetailsCacheImpl(contractDetails));
+    /**
+     * Loads the state of the account into <b>this object' caches</b>.
+     * Requires write locks on both {@link #lockAccounts} and {@link #lockDetails}.
+     *
+     * @implNote If the calling method has acquired a weaker lock, the lock must be released before calling this method.
+     * @apiNote If the account was never stored this call will create it.
+     */
+    private void loadAccountState(Address address) {
+        lockAccounts.writeLock().lock();
+        lockDetails.writeLock().lock();
+        try {
+            repository.loadAccountState(address, this.cachedAccounts, this.cachedDetails);
+        } finally {
+            lockAccounts.writeLock().unlock();
+            lockDetails.writeLock().unlock();
         }
     }
 
     @Override
-    public synchronized void deleteAccount(Address address) {
-        getAccountState(address).delete();
-        getContractDetails(address).setDeleted(true);
+    public void deleteAccount(Address address) {
+        lockAccounts.writeLock().lock();
+        lockDetails.writeLock().lock();
+        try {
+            getAccountState(address).delete();
```

Comment:
My impression is that this wouldn't work, due to getAccountState() requiring a ``readLock``.

Reply:
Nevermind I am incorrect on this point.